### DISCUSSION TITLE
What to unit test in remix.run?

### DISCUSSION THREAD

**Discussion by otatar** [remix-run/remix#8381] [Q&A]
Hi there,

I've just fell in love with remix.run and its documentation is clear. But on thing is quite blurry, and that is testing. 
So what to test in remix.run app?
How to test loader and actions? Like this:  [https://sergiodxa.com/articles/test-remix-loaders-and-actions](url)
Or using `createRemixStub()`?
jest or vitetest

**Comment by sergiodxa** [comment]
`createRemixStub` from `@remix-run/testing` package is for when you want to test a component that depends on Remix, e.g. a component rendering a <Form> or a <Link>, or a component using hook like `useFetcher`, `useSubmit`, `useSearchParams`, etc. The `createRemixStub` function will help you simulate the component is rendered inside Remix and let you provider loaders and actions to test interacting with the component works.

If you only want to test loaders and actions, it's way simpler to just call them like normal functions, you don't need to mock a UI to test them.

**Comment by otatar** [comment]
Thank you Sergio...
So here is my strategy for testing Remix app:
Unit testing:
  a) Test loader/actions like normal functions
  b) Test components using `createRemixStub` and providing mocked loaders and actions
   **note: It looks like you must have two test files form this two cases (a) and b)) because loader/actions must be tested in `node` (server) environment, and component testing must be done in `jsdom` (browser) environment. I haven't managed to switch between environment in single test file

End-to-End testing: Using Cypress on Playwright on mocked DB

Can I do better that this?
